xray/commands/curation/analyse.go

```diff
@@ -0,0 +1,170 @@
+package curation
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/jfrog/jfrog-cli-core/v2/utils/coreutils"
+	utils2 "github.com/jfrog/jfrog-cli-core/v2/xray/commands/utils"
+	"github.com/jfrog/jfrog-client-go/artifactory"
+	"github.com/jfrog/jfrog-client-go/auth"
+	"github.com/jfrog/jfrog-client-go/utils/io/httputils"
+	"github.com/jfrog/jfrog-client-go/utils/log"
+	"github.com/jfrog/jfrog-client-go/xray/services/utils"
+	"net/http"
+	"strings"
+)
+
+type treeAnalyzer struct {
+	artiManager       artifactory.ArtifactoryServicesManager
+	artAuth           auth.ServiceDetails
+	httpClientDetails httputils.HttpClientDetails
+	url               string
+	repo              string
+	tech              coreutils.Technology
+}
+
+const (
+	BlockMessageKey  = "jfrog packages curation"
+	NotBeingFoundKey = "not being found"
+)
+
+func (nc *treeAnalyzer) recursiveNodeCuration(graph *utils.GraphNode, respStatus *[]PackageStatus, parent, parentVersion string, isRoot bool) error {
+	if parent == "" && !isRoot {
+		_, parent, parentVersion = getUrlNameAndVersionByTech(nc.tech, graph.Id, nc.url, nc.repo)
+	}
+	for _, node := range graph.Nodes {
+		packageUrl, name, version := getUrlNameAndVersionByTech(nc.tech, node.Id, nc.url, nc.repo)
+		resp, _, err := nc.artiManager.Client().SendHead(packageUrl, &nc.httpClientDetails)
+		if err != nil || resp.StatusCode >= 400 {
+			if resp == nil {
+				return err
+			}
+			if resp.StatusCode != http.StatusForbidden {
+				log.Error(fmt.Sprintf("Failed to head package %s for download url: %s, status-code: %v, err: %v", node.Id, packageUrl, resp.StatusCode, err))
```

Comment:
Our approach is to avoid `log.Error` when possible, and instead accumulate the error messages, so that they can be returned at the end of the loop as part of an error. 
The same goes for the other log.Error in this function. Also, the other error message is missing some content. Let's implement the same in the `addBlockedPackage` function.

Failed to head package %s for download url: %s, status-code: %v, err: %v", node.Id, packageUrl, resp.StatusCode, err
-->
Failed sending HEAD request to %s for package '%s'. Status code: %v. Cause: %v", packageUrl, node.Id, resp.StatusCode, err